src/tree.py

```diff
@@ -0,0 +1,124 @@
+from typing import Dict, List
+from src.datapoint import DataSet, Gini
+import numpy as np
+
+
+class Node:
+    def __init__(self, datapoint: DataSet):
+        """
+        Represents a Node of the Tree
+        :param datapoint: Datapoint object
+        """
+        self.data_node = datapoint
+        self._pointer: Pointer = Pointer()
+
+    # def __repr__(self) -> str:
+    #     if len(self._pointer.list_node):
+    #         return f'{self.data_node} -> {self._pointer}'
+    #     else:
+    #         return f'{self.data_node}'
+
+    @property
+    def get_data(self) -> DataSet:
+        return self.data_node
+
+    def add_node_to_pointer(self, node) -> None:
+        self._pointer.list_node.append(node)
+
+    # def find_pointer(self, data_node) -> "Pointer":
+        # return next(pointer for pointer in self._pointer if pointer.get_pointer == data_node)
+
+
+class Pointer:
+    def __init(self):
+        self.list_node: List[Node] = []
+        self._data_node = None
+        self._node = None
+
+    def __repr(self):
+        return f'{self._data_node} -> {self._node}'
+
+    @property
+    def get_pointer(self) -> List[Node]:
+        return self.list_node
+
+
+class Leaf:
+    def __init__(self, dataset: DataSet):
+        self._dataset = dataset
+
+
+class Tree:
+    def __init__(self):
+        self._exit: bool = False
+        self._list_node = []
+
+    def _add_node(self, node):
+        self._list_node.append(node)
+
+    def _exit_condition(self, node):
```

Comment:
Rename this to be explicit
```suggestion
    def _is_dataset_pure(self, data):
```